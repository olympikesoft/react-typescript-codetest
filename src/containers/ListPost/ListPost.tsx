// PostList.jsx
import React, { useEffect, useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { fetchPosts } from '../../store/actions/Posts/fetchPost';
import { selectPost } from '../../store/actions/Posts/selectPost';
import { selectPosts, selectPostsLoading, selectPostsError } from '../../store/selectors/postsSelectors';
import PostCard from '../../components/Post/PostCard/PostCard';
import { AppDispatch } from '../../store/store';
import { SearchInput } from '../../components/Input/SearchInput/SearchInput';
import Grid from '@mui/material/Grid';
import { selectSearchTerm } from '../../store/selectors/searchSelectors';
import { selectFilteredPosts } from '../../store/selectors/searchFilteredPostSelector';
import { Post, searchFilterState } from '../../interfaces';
import { selectUsers } from '../../store/selectors/usersSelectors';
import { fetchUsers } from '../../store/actions/Users/fetchUsers';

const ListPost: React.FC = () => {
  const dispatch = useDispatch<AppDispatch>();
  const posts = useSelector(selectPosts);
  const users = useSelector(selectUsers);
  const loading = useSelector(selectPostsLoading);
  const error = useSelector(selectPostsError);
  const searchTerm = useSelector(selectSearchTerm) || '';
  const selectFilter:searchFilterState  = {
    users: users,
    posts: posts,
  }
  const filteredPosts = useSelector(_state => selectFilteredPosts(selectFilter, searchTerm));

  const [selectedPost, setSelectedPost] = useState<number>(-1);

  useEffect(() => {
    dispatch(fetchPosts());
    dispatch(fetchUsers());
  }, [dispatch]);

  if (loading) return <p>Loading...</p>;
  if (error) return <p>Error loading posts!</p>;

  const onSelectPost = (post: Post) =>{
    setSelectedPost(post.id)
    dispatch(selectPost(post));
  }

  return (
    <div>
      <Grid container direction="column" spacing={2}>
        <Grid item>
          <SearchInput />
        </Grid>
        <Grid item>    
          {filteredPosts.map(post => {
            const postWithDefaultInfo = {
              ...post,
              createdAt: post.createdAt || new Date(),
              userImage: post.userImage || 'https://placehold.co/64x64'
            };

            return (
              <PostCard
                key={post.id}
                post={postWithDefaultInfo}
                onSelect={onSelectPost}
                isSelected={post.id === selectedPost}
              />
            );
          })}
        </Grid>
      </Grid>
    </div>
  );
};

export default ListPost;
