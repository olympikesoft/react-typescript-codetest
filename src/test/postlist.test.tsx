/* eslint-disable testing-library/prefer-screen-queries */
/* eslint-disable testing-library/no-debugging-utils */
import React from 'react';
import { render } from '@testing-library/react';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import ListPost from '../containers/ListPost/ListPost';
import thunk from 'redux-thunk';
import { screen } from '@testing-library/react';
import '@testing-library/jest-dom';


const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const mockPosts = [
  { id: 1, title: 'Post 1', body: 'This is a mock post', userId: 1 },
];

const mockUsers = [
  { id: 1, name: 'User 1', username: 'user1', email: 'user1@example.com' },
];

const initialState = {
  posts: {
    data: mockPosts, 
    loading: false,
    error: null
  },
  users: {
    data: mockUsers, 
    loading: false,
    error: null
  },
};

const store = mockStore(initialState);

describe('InitialPage', () => {
  it('renders without crashing and displays posts', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <ListPost />
      </Provider>
    );

    const searchInput = screen.getByLabelText(/search posts/i);
    expect(searchInput).toBeInTheDocument();


    expect(asFragment()).toMatchSnapshot();
  });
});
