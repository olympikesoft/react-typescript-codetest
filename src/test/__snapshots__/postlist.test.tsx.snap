// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InitialPage renders without crashing and displays posts 1`] = `
<DocumentFragment>
  <div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-direction-xs-column css-1rl1og4-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for=":r0:"
            id=":r0:-label"
          >
            Search Posts
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id=":r0:"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-yjsfm1"
              >
                <span>
                  Search Posts
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      />
    </div>
  </div>
</DocumentFragment>
`;
